### PR TITLE
use normal impersonation headers to speaking back to ourselves

### PR DESCRIPTION
**Pull request by deads2k**
I think this is right, but callers need to be chased to be sure its for loopbacks.  I suspect some of these can be made private.

/assign @simo5 
/assign @enj

### CODE REVIEW

File: pkg/oauthserver/client/impersonate.go [hunk 1]
```diff
@@ -68,9 +43,11 @@ func NewImpersonatingRESTClient(user user.Info, client restclient.Interface) res
 
 // Verb does the impersonation per request by setting the proper headers
 func (c impersonatingRESTClient) impersonate(req *restclient.Request) *restclient.Request {
-	req.SetHeader(authenticationapi.ImpersonateUserHeader, c.user.GetName())
-	req.SetHeader(authenticationapi.ImpersonateGroupHeader, c.user.GetGroups()...)
-	req.SetHeader(authenticationapi.ImpersonateUserScopeHeader, c.user.GetExtra()[authorizationapi.ScopesKey]...)
+	req.SetHeader(transport.ImpersonateUserHeader, c.user.GetName())
+	req.SetHeader(transport.ImpersonateGroupHeader, c.user.GetGroups()...)
+	for k, vv := range c.user.GetExtra() {
+		req.SetHeader(transport.ImpersonateUserExtraHeaderPrefix+k, vv...)
```
Comment (deads2k) [inline pkg/oauthserver/client/impersonate.go]: This is a different header than scopes, but should be respected for several releases of servers now

**Comment by simo5** [comment]
Are you proposing this instead of #18378 ?
What's wrong with that approach ?

File: pkg/oauthserver/client/impersonate.go (hunk 1 above)
Comment (simo5) [reply pkg/oauthserver/client/impersonate.go]: Do we really need to put in all Extras when impersonating ?

File: pkg/oauthserver/client/impersonate.go [hunk 2]
```diff
@@ -5,47 +5,22 @@ import (
...
-	}
-	return rt.delegate.RoundTrip(req)
-}
-
 // NewImpersonatingConfig wraps the config's transport to impersonate a user, including user, groups, and scopes
 func NewImpersonatingConfig(user user.Info, config restclient.Config) restclient.Config {
 	oldWrapTransport := config.WrapTransport
 	config.WrapTransport = func(rt http.RoundTripper) http.RoundTripper {
-		return newImpersonatingRoundTripper(user, oldWrapTransport(rt))
+		return transport.NewImpersonatingRoundTripper(transport.ImpersonationConfig{
```
Comment (simo5) [inline pkg/oauthserver/client/impersonate.go]: Our call used to delete all headers, looking upstream tough this is not done in their RoundTrip function, groups and extras are Added to existing headers. Doesn't this risk incorrect impersonation by adding more privileged groups scopes ?

Comment (simo5) [reply pkg/oauthserver/client/impersonate.go]: Fundamentally the difference here is that if there is already impersonation, new impersonation is skipped now, is this correct ?

File: pkg/oauthserver/client/impersonate.go (hunk 1 above)
Comment (deads2k) [reply pkg/oauthserver/client/impersonate.go]: > Do we really need to put in all Extras when impersonating ?

for the general case, yes

File: pkg/oauthserver/client/impersonate.go (hunk 2 above)
Comment (deads2k) [reply pkg/oauthserver/client/impersonate.go]: > Fundamentally the difference here is that if there is already impersonation, new impersonation is skipped now, is this correct ?

Correct.  I don't think any existing callers were hoping for double impersonation powers, but I could see an argument to change upstream if you wish.

**Comment by deads2k** [comment]
I'd like to collapse onto the upstream.  closing since https://github.com/openshift/origin/pull/18378 is close

**Pull request closed without merging**